### PR TITLE
Cleanup MobileCore APIs

### PR DESCRIPTION
**Pull request by praveek**
Cleanup MobileCore APIs to use the new Eventhub. 

#### New API

Added new API so that extensions can explicitly specify timeout for each call. 
`public static void dispatchEventWithResponseCallback(final Event event, long timeoutMS, final AdobeCallbackWithError<Event> responseCallback)`

#### Deprecated APIs

These APIs were only used by Extensions. Going forward, they can use equivalent APIs in ExtensionAPI class. 
`public static void dispatchEventWithResponseCallback(final Event event,
                                                         final AdobeCallbackWithError<Event> responseCallback)`

`public static boolean dispatchEvent(final Event event, final ExtensionErrorCallback<ExtensionError> errorCallback)`

`public static boolean dispatchEventWithResponseCallback(final Event event,
                                                            final AdobeCallback<Event> responseCallback,
                                                            final ExtensionErrorCallback<ExtensionError> errorCallback)`

`public static boolean dispatchResponseEvent(final Event responseEvent, final Event requestEvent,
                                                final ExtensionErrorCallback<ExtensionError> errorCallback)`

Added seperate issues to fix
1) [Cleanup App class used to store ApplicationContext](https://github.com/adobe/aepsdk-core-android/issues/151)
2) [Remove MobileCore.getEventHistory API. ](https://github.com/adobe/aepsdk-core-android/issues/150)

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/eventhub/EventHubExtensionRegistarar.kt [hunk 1]
```diff
@@ -0,0 +1,90 @@
+package com.adobe.marketing.mobile.internal.eventhub
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/eventhub/EventHubExtensionRegistarar.kt:1]: Missing copyright and license headers.

File: code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/eventhub/EventHubExtensionRegistarar.kt [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+    private val extensionsRegisteredBeforeStart = mutableListOf<Class<out Extension>>()
+
+    fun registerExtension(
+        extensionClass: Class<out Extension>,
+        errorCallback: ExtensionErrorCallback<ExtensionError>?,
+    ): Boolean {
+        synchronized(mutex) {
+            if (!startCalled) {
+                extensionsRegisteredBeforeStart.add(extensionClass)
+            }
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/eventhub/EventHubExtensionRegistarar.kt:28]: Should we return and log a warning message if SDK was started?

**Comment by yangyansong-adbe** [comment]
IMO, the `pendingHubStart` class is way too complicated. Can we just use an atomic boolean and a concurrent list to guard the `Eventhub start` and the `register extension` operations? If the Eventhub was started the later start/registerExtension calls should be ignored. If the extension registration failed,  we don't even need to remove it from the registered extension list as we don't support registering the same extension multiple times.

**Comment by praveek** [comment]
I agree. I added the logic to support current registration logic. I have removed `EventHubExtensionRegistarar` class for now and will revisit as part of https://github.com/adobe/aepsdk-core-android/issues/155

File: code/android-core-library/src/main/java/com/adobe/marketing/mobile/Extension.java [hunk 3]
```diff
@@ -58,7 +58,7 @@ protected String getVersion() {
 	/**
 	 * Optional metadata provided for use by the event hub for logging.
 	 *
-	 * @return the extension metadata as a {@link Map<String, String}
+	 * @return the extension metadata as a {@code Map<String, Object>}
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/main/java/com/adobe/marketing/mobile/Extension.java:61]: The `getMetadata` method returns an instance of `Map<String, String>`.

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 4]
```diff
@@ -250,118 +193,103 @@ public static void configureWithAppID(final String appId) {
...
-     * @param errorCallback optional {@link ExtensionErrorCallback} which will be called if an error occurred during dispatching
-     * @return {@code boolean} indicating if the the event dispatching operation succeeded
+     * @param completionCallback An optional {@link AdobeCallback} invoked after registrations are completed
      */
-    public static boolean dispatchEvent(final Event event, final ExtensionErrorCallback<ExtensionError> errorCallback) {
-        if (eventHub == null) {
-            Log.debug(LOG_TAG, "Failed to dispatch event. (%s)", NULL_CONTEXT_MESSAGE);
+    public static void start(final AdobeCallback<?> completionCallback) {
+        if(!sdkInitializedWithContext.get()) {
+            Log.error(LOG_TAG, "Failed to start - completionCallback is null");
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:239]: We need to update this error log.

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 5]
```diff
@@ -373,153 +301,227 @@ public void call(final Event e) {
...
+     * Configure the SDK by downloading the remote configuration file hosted on Adobe servers
+     * specified by the given application ID.
+     * <p>
+     * The configuration file is cached once downloaded and used in subsequent calls to this
+     * API. If the remote file is updated after the first download, the updated file is downloaded
+     * and replaces the cached file.
+     * @param appId A unique identifier assigned to the app instance by Adobe Launch
+     */
+    public static void configureWithAppID(final String appId) {
+        Map<String, Object> eventData = new HashMap<>();
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:481]: Can we add a null/empty check here?

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 6]
```diff
@@ -373,153 +301,227 @@ public void call(final Event e) {
...
+     * Calls to this API will replace any existing SDK configuration except those set using
+     * {@link #updateConfiguration(Map)} or {@link #setPrivacyStatus(MobilePrivacyStatus)}.
+     * Configuration updates made using {@link #updateConfiguration(Map)} and
+     * {@link #setPrivacyStatus(MobilePrivacyStatus)} are always applied on top of configuration changes
+     * made using this API.
+     *
+     * @param fileName the name of the configure file in the assets folder. A value of {@code null} has no effect.
+     */
+    public static void configureWithFileInAssets(final String fileName) {
+        Map<String, Object> eventData = new HashMap<>();
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:509]: null/empty check?

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 7]
```diff
@@ -529,21 +531,13 @@ public static void configureWithFileInAssets(final String fileName) {
...
-        if (eventHub == null) {
-            Log.debug(LOG_TAG, "Failed to load configuration with file path (%s)", NULL_CONTEXT_MESSAGE);
-            return;
-        }
-
-        if (StringUtils.isNullOrEmpty(filepath)) {
-            Log.warning("Configuration", "Unable to configure with null or empty file path");
-            return;
-        }
+        Map<String, Object> eventData = new HashMap<>();
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:534]: null/empty check?

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 8]
```diff
@@ -559,19 +553,13 @@ public static void configureWithFileInPath(final String filepath) {
...
-        // Create a EventData Map <String, Object>
-        HashMap<String, Variant> eventDataMap = new HashMap<String, Variant>();
-        Variant configMapVariant = Variant.fromTypedMap(configMap, PermissiveVariantSerializer.DEFAULT_INSTANCE);
-        eventDataMap.put(CoreConstants.EventDataKeys.Configuration.CONFIGURATION_REQUEST_CONTENT_UPDATE_CONFIG,
-                configMapVariant);
-        EventData eventData = new EventData(eventDataMap);
-        final Event event = new Event.Builder("Configuration Update", EventType.CONFIGURATION,
-                EventSource.REQUEST_CONTENT).setData(eventData).build();
-        eventHub.dispatch(event);
+        Map<String, Object> eventData = new HashMap<>();
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:556]: null/empty check?

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 9]
```diff
@@ -846,195 +729,201 @@ public static void lifecycleStart(final Map<String, String> additionalContextDat
...
-            Log.debug(LOG_TAG, "Could not trigger PII, the data is null or empty.");
-            return;
-        }
-
-        final EventData eventData = new EventData()
-                .putStringMap(CoreConstants.EventDataKeys.Signal.SIGNAL_CONTEXT_DATA, data);
-        eventHub.dispatch(new Event.Builder("CollectPII", EventType.GENERIC_PII,
-                EventSource.REQUEST_CONTENT).setData(eventData).build());
-        Log.trace(LOG_TAG, "Collect PII event was sent");
+    public static void trackAction(final String action, final Map<String, String> contextData) {
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:757]: null/empty check?

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 10]
```diff
@@ -846,195 +729,201 @@ public static void lifecycleStart(final Map<String, String> additionalContextDat
...
+     * <p>
      *
-     * @param resourceID the resource Id of the icon
+     * @param state       {@code String} containing the name of the state to track
+     * @param contextData contextData {@code Map<String, String>} containing context data to attach on this hit
      */
-    public static void setSmallIconResourceID(final int resourceID) {
-        App.setSmallIconResourceID(resourceID);
+    public static void trackState(final String state, final Map<String, String> contextData) {
+        Map<String, Object> eventData = new HashMap<>();
```
Comment (yangyansong-adbe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:778]: null/empty check?

**Review by yangyansong-adbe — APPROVED** [review]
LGTM, only several small comments.

**Comment by yangyansong-adbe** [comment]
[Improve code inspection with annotations 
](https://developer.android.com/studio/write/annotations)
@praveek one more question, can we consider adding `@Nullable`/`@NonNull` annotations to the public APIs in SDK 2.0?

File: code/android-core-library/src/main/java/com/adobe/marketing/mobile/EventType.java [hunk 11]
```diff
@@ -27,6 +27,12 @@ public final class EventType {
 	// Todo - Expose String constants. Remove 'TYPE' prefix after fixing build issues
 	public static final String TYPE_WILDCARD = "com.adobe.eventType._wildcard_";
 	public static final String TYPE_HUB = "com.adobe.eventType.hub";
+	public static final String TYPE_CONFIGURATION = "com.adobe.eventType.configuration";
+	public static final String TYPE_GENERIC_LIFECYLE = "com.adobe.eventType.generic.lifecycle";
```
Comment (spoorthipujariadobe) [inline code/android-core-library/src/main/java/com/adobe/marketing/mobile/EventType.java:31]: Typo here. Should be `TYPE_GENERIC_LIFECYCLE`

File: code/android-core-library/src/main/java/com/adobe/marketing/mobile/Extension.java (hunk 3 above)
Comment (spoorthipujariadobe) [inline code/android-core-library/src/main/java/com/adobe/marketing/mobile/Extension.java:61]: Return type is Map<String, String>, typo in javadocs

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 12]
```diff
@@ -250,118 +193,103 @@ public static void configureWithAppID(final String appId) {
...
         }
 
-        eventHub.dispatch(event);
-        return true;
+        EventHub.Companion.getShared().registerListener(eventType, eventSource, callback);
     }
 
     /**
      * This method will be used when the provided {@code Event} is used as a trigger and a response event
      * is expected in return. The returned event needs to be sent using
```
Comment (spoorthipujariadobe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:277]: Need to update javadocs here. `dispatchEvent` does not expect a response event in return right?

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java [hunk 13]
```diff
@@ -580,17 +568,13 @@ public static void updateConfiguration(final Map<String, Object> configMap) {
...
-        EventData eventData = new EventData();
-        eventData.putBoolean(CoreConstants.EventDataKeys.Configuration.CONFIGURATION_REQUEST_CONTENT_CLEAR_UPDATED_CONFIG,
+        Map<String, Object> eventData = new HashMap<>();
+        eventData.put(CoreConstants.EventDataKeys.Configuration.CONFIGURATION_REQUEST_CONTENT_CLEAR_UPDATED_CONFIG,
                 true);
-        final Event event = new Event.Builder("Clear updated configuration", EventType.CONFIGURATION,
-                EventSource.REQUEST_CONTENT).setData(eventData).build();
-        eventHub.dispatch(event);
+
+        Event event = new Event.Builder("Configuration Update", EventType.CONFIGURATION,
```
Comment (spoorthipujariadobe) [inline code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:575]: Event name is changed, need to update `MobileCoreTests` if we want to use the new event name

Comment (spoorthipujariadobe) [reply code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:575]: Nvm, just saw you have changed the tests

**Review by spoorthipujariadobe — APPROVED** [review]

File: code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java (hunk 9 above)
Comment (praveek) [reply code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java]: I will check with the Analytics squad and update trackState/trackAction later

**Pull request merged**